File: go/623.go
package main

// Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func right(t *TreeNode, v, d int) *TreeNode { // for right subtree
	if d == 1 { // target depth
		var ret *TreeNode = &TreeNode{Val: v, Left: nil, Right: t}
		return ret
	}
	if t == nil { // under leaf, but not reach target depth
		return nil
	}
	d-- // update height
	t.Left = left(t.Left, v, d) // recursive
	t.Right = right(t.Right, v, d) // recursive
	return t
}

func left(t *TreeNode, v, d int) *TreeNode { // for left subtree
	if d == 1 {
		var ret *TreeNode = &TreeNode{Val: v, Left: t, Right: nil}
		return ret
	}
	if t == nil {
		return nil
	}
	d--
	t.Left = left(t.Left, v, d)
	t.Right = right(t.Right, v, d)
	return t
}

func addOneRow(root *TreeNode, val int, depth int) *TreeNode {
	return left(root, val, depth) // treat root as left subtree case
}

File: cpp/623.cpp
#include <iostream>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};

class Solution {
public:
  TreeNode *right(TreeNode *t, int v, int d) { // right subtree
    if (d == 1) {                              // target depth
      TreeNode *l = new TreeNode(v, nullptr, t);
      return l;
    }
    if (!t) // leaf but not reach target depth
      return nullptr;
    d--;                              // update depth
    t->left = left(t->left, v, d);    // recursive
    t->right = right(t->right, v, d); // recursive
    return t;
  }
  TreeNode *left(TreeNode *t, int v, int d) {
    if (d == 1) {
      TreeNode *r = new TreeNode(v, t, nullptr);
      return r;
    }
    if (!t)
      return nullptr;
    d--;
    t->left = left(t->left, v, d);
    t->right = right(t->right, v, d);
    return t;
  }
  TreeNode *addOneRow(TreeNode *root, int val, int depth) {
    return left(root, val, depth); // treat root as left subtree case
  }
};
